tools/run_tests/run_xds_tests.py

```diff
@@ -499,6 +535,93 @@ def test_secondary_locality_gets_requests_on_primary_failure(
+    wait_for_healthy_backends(gcp, original_backend_service, instance_group)
+
+    patch_backend_instances(gcp, alternate_backend_service,
+                            [same_zone_instance_group])
+    logger.info('waiting for alternate to become healthy')
+    wait_for_healthy_backends(gcp, alternate_backend_service,
+                              same_zone_instance_group)
+
+    original_backend_instances = get_instance_names(gcp, instance_group)
+    logger.info('original backends instances: %s', original_backend_instances)
+
+    alternate_backend_instances = get_instance_names(gcp,
+                                                     same_zone_instance_group)
+    logger.info('alternate backends instances: %s', alternate_backend_instances)
+
+    # Start with all traffic going to original_backend_service.
+    logger.info('waiting for traffic to all go to original')
+    wait_until_all_rpcs_go_to_given_backends(original_backend_instances,
+                                             _WAIT_FOR_STATS_SEC)
+
+    try:
+        # Path urlmap, change route action to traffic splitting between original
+        # and alternate.
+        logger.info('patching url map with traffic splitting')
+        expected_service_percentage = [20, 80]
+        patch_url_map_weighted_backend_services(
+            gcp, {
+                original_backend_service: expected_service_percentage[0],
+                alternate_backend_service: expected_service_percentage[1],
+            })
+        expected_instance_percentage = [
+            expected_service_percentage[0] * 1.0 /
+            len(original_backend_instances)
+        ] * len(original_backend_instances) + [
+            expected_service_percentage[1] * 1.0 /
+            len(alternate_backend_instances)
+        ] * len(alternate_backend_instances)
+
+        # Wait for traffic to go to both services.
+        logger.info(
+            'waiting for traffic to go to all backends (including alternate)')
+        wait_until_all_rpcs_go_to_given_backends(
+            original_backend_instances + alternate_backend_instances,
+            _WAIT_FOR_STATS_SEC)
+
+        # Verify that weights between two services are expected.
+        retry_count = 3
+        for i in range(retry_count):
+            stats = get_client_stats(_NUM_TEST_RPCS, _WAIT_FOR_STATS_SEC)
+            got_instance_count = [
+                stats.rpcs_by_peer[i] for i in original_backend_instances
+            ] + [stats.rpcs_by_peer[i] for i in alternate_backend_instances]
+            total_count = sum(got_instance_count)
+            got_instance_percentage = [
+                x * 100.0 / total_count for x in got_instance_count
+            ]
+
+            try:
+                compareDistributions(got_instance_percentage,
+                                     expected_instance_percentage, 5)
+            except Exception as e:
+                logger.warning('attempt %d', i)
```

Comment:
Warning seems too strong here since we probably expect the propagation of the new weights to take more than 10 seconds fairly frequently, meaning this block will execute even during a successful test run..